File: Backend/src/main/java/edu/ntnu/idatt2106/krisefikser/service/storage/StorageService.java
package edu.ntnu.idatt2106.krisefikser.service.storage;

import edu.ntnu.idatt2106.krisefikser.api.dto.item.ItemResponseDto;
import edu.ntnu.idatt2106.krisefikser.api.dto.storage.StorageItemResponseDto;
import edu.ntnu.idatt2106.krisefikser.persistance.entity.household.Household;
import edu.ntnu.idatt2106.krisefikser.persistance.entity.item.Item;
import edu.ntnu.idatt2106.krisefikser.persistance.entity.storage.StorageItem;
import edu.ntnu.idatt2106.krisefikser.persistance.entity.user.User;
import edu.ntnu.idatt2106.krisefikser.persistance.enums.ItemType;
import edu.ntnu.idatt2106.krisefikser.persistance.repository.household.HouseholdRepository;
import edu.ntnu.idatt2106.krisefikser.persistance.repository.item.ItemRepository;
import edu.ntnu.idatt2106.krisefikser.persistance.repository.storage.StorageItemRepository;
import edu.ntnu.idatt2106.krisefikser.persistance.repository.user.UserRepository;
import java.time.LocalDateTime;
import java.util.List;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * Service class for managing storage items in a household. This includes adding, removing,
 * updating, and retrieving storage items. It also provides methods to filter items by type and
 * expiration date.
 */
@Service
public class StorageService {

  private static final Logger logger = LoggerFactory.getLogger(StorageService.class);

  private final StorageItemRepository storageItemRepository;
  private final HouseholdRepository householdRepository;
  private final ItemRepository itemRepository;
  private final UserRepository userRepository;

  /**
   * Constructor for StorageService.
   *
   * @param storageItemRepository The repository for storage item operations.
   * @param householdRepository   The repository for household operations.
   * @param itemRepository        The repository for item operations.
   */
  public StorageService(StorageItemRepository storageItemRepository,
      HouseholdRepository householdRepository,
      ItemRepository itemRepository, UserRepository userRepository) {
    this.storageItemRepository = storageItemRepository;
    this.householdRepository = householdRepository;
    this.itemRepository = itemRepository;
    logger.info("StorageService instantiated.");
    this.userRepository = userRepository;
  }

  /**
   * Get all storage items for a specific household.
   *
   * @return A list of StorageItemResponseDto.
   */
  public List<StorageItemResponseDto> getStorageItemsByHousehold() {
    logger.info("Fetching storage items for current household");
    Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
    String email = authentication.getName();
    User user = userRepository.findByEmail(email)
        .orElseThrow(() -> new IllegalArgumentException("No user logged in"));
    Household household = user.getHousehold();
    List<StorageItemResponseDto> items = storageItemRepository
        .findByHouseholdId(household.getId())
        .stream()
        .map(storageItem -> {
          logger.debug("Mapping StorageItem id={} to DTO", storageItem.getId());
          return new StorageItemResponseDto(
              storageItem.getId(),
              new ItemResponseDto(
                  storageItem.getItem().getId(),
                  storageItem.getItem().getName(),
                  storageItem.getItem().getCaloricAmount(),
                  storageItem.getItem().getItemType()
              ),
              storageItem.getHousehold().getId(),
              storageItem.getUnit(),
              storageItem.getAmount(),
              storageItem.getExpirationDate()
          );
        })
        .toList();
    logger.info("Found {} storage items for householdId={}", items.size(), household.getId());
    return items;
  }

  /**
   * Get storage items for a specific household filtered by item type.
   *
   * @param itemType The type of items to filter by.
   * @return A list of StorageItem entities.
   */
  public List<StorageItem> getStorageItemsByHouseholdAndType(ItemType itemType) {
    logger.info("Fetching storage items for with itemType={}",
        itemType);
    Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
    String email = authentication.getName();
    Household household = userRepository.findByEmail(email)
        .orElseThrow(() -> new IllegalArgumentException("No user logged in"))
        .getHousehold();

    List<StorageItem> items =
        storageItemRepository.findByHouseholdIdAndItemItemType(household.getId(),
            itemType);
    logger.info("Found {} items of type {} for householdId={}", items.size(), itemType,
        household.getId());
    return items;
  }

  /**
   * Get items that will expire before a specific date.
   *
   * @param before The date before which items will expire.
   * @return A list of StorageItem entities that will expire before the specified date.
   */
  public List<StorageItem> getExpiringItems(LocalDateTime before) {
    logger.info("Fetching items expiring before {}", before);
    Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
    String email = authentication.getName();
    User user = userRepository.findByEmail(email)
        .orElseThrow(() -> new IllegalArgumentException("No user logged in"));
    Household household = user.getHousehold();

    List<StorageItem> items = storageItemRepository.findByHouseholdIdAndExpirationDateBefore(
        household.getId(), before);
    logger.info("Found {} expiring items for householdId={}", items.size(), household.getId());
    return items;
  }

  /**
   * Get items that have already expired.
   *
   * @param householdId The ID of the household.
   * @return A list of StorageItem entities that have already expired.
   */
  public List<StorageItem> getExpiredItems(String householdId) {
    LocalDateTime now = LocalDateTime.now();
    logger.info("Fetching items already expired as of {} for householdId={}", now, householdId);
    List<StorageItem> items = storageItemRepository.findByHouseholdIdAndExpirationDateBefore(
        householdId, now);
    logger.info("Found {} expired items for householdId={}", items.size(), householdId);
    return items;
  }

  /**
   * Adds an item to the storage of a household.
   *
   * @param itemId         The ID of the item to be added.
   * @param unit           The unit of measurement for the item.
   * @param amount         The amount of the item to be added.
   * @param expirationDate The expiration date of the item.
   * @return The newly created StorageItem entity.
   */
  @Transactional
  public StorageItem addItemToStorage(Long itemId,
      String unit, Integer amount,
      LocalDateTime expirationDate) {
    logger.info(
        "Adding item to storage: itemId={}, amount={}, unit={}, expiration={}",
        itemId, amount, unit, expirationDate);

    // Get the currently authenticated user
    Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
    String email = authentication.getName();
    User user = userRepository.findByEmail(email)
        .orElseThrow(() -> new IllegalArgumentException("No user logged in"));
    Household household = user.getHousehold();

    Item item = itemRepository.findById(itemId)
        .orElseThrow(() -> {
          logger.error("Item not found: {}", itemId);
          return new IllegalArgumentException("Item not found");
        });

    StorageItem storageItem = new StorageItem();
    storageItem.setHousehold(household);
    storageItem.setItem(item);
    storageItem.setUnit(unit);
    storageItem.setAmount(amount);
    storageItem.setExpirationDate(expirationDate);
    storageItem.setDateAdded(LocalDateTime.now());

    StorageItem saved = storageItemRepository.save(storageItem);
    logger.info("StorageItem created with id={}", saved.getId());
    return saved;
  }

  /**
   * Removes an item from storage.
   *
   * @param storageItemId The ID of the storage item to be removed.
   */
  @Transactional
  public void removeItemFromStorage(Long storageItemId) {
    logger.info("Removing storage item with id={}", storageItemId);
    storageItemRepository.deleteById(storageItemId);
    logger.info("Storage item {} removed", storageItemId);
  }

  /**
   * Updates a storage item with new values.
   *
   * @param storageItemId  The ID of the storage item to update.
   * @param unit           The new unit of measurement.
   * @param amount         The new amount.
   * @param expirationDate The new expiration date.
   * @return The updated StorageItem entity.
   */
  @Transactional
  public StorageItem updateStorageItem(Long storageItemId, String unit, Integer amount,
      LocalDateTime expirationDate) {
    logger.info("Updating storage item id={} with unit={}, amount={}, expiration={}",
        storageItemId, unit, amount, expirationDate);

    StorageItem storageItem = storageItemRepository.findById(storageItemId)
        .orElseThrow(() -> {
          logger.error("Storage item not found: {}", storageItemId);
          return new IllegalArgumentException("Storage item not found");
        });

    if (unit != null) {
      storageItem.setUnit(unit);
      logger.debug(" - unit set to {}", unit);
    }

    if (amount != null) {
      storageItem.setAmount(amount);
      logger.debug(" - amount set to {}", amount);
    }

    storageItem.setExpirationDate(expirationDate);
    logger.debug(" - expirationDate set to {}", expirationDate);

    StorageItem updated = storageItemRepository.save(storageItem);
    logger.info("Storage item {} updated successfully", storageItemId);
    return updated;
  }

  /**
   * Updates the amount of a storage item in the household's storage.
   *
   * @param storageItemId The ID of the storage item to be updated.
   * @param newAmount     The new amount to set for the storage item.
   * @return The updated StorageItem entity.
   */
  @Transactional
  public StorageItem updateItemAmount(Long storageItemId, Integer newAmount) {
    logger.info("Updating amount of storage item id={} to {}", storageItemId, newAmount);

    StorageItem storageItem = storageItemRepository.findById(storageItemId)
        .orElseThrow(() -> {
          logger.error("Storage item not found: {}", storageItemId);
          return new IllegalArgumentException("Storage item not found");
        });

    storageItem.setAmount(newAmount);
    StorageItem saved = storageItemRepository.save(storageItem);
    logger.info("Storage item {} amount updated to {}", storageItemId, newAmount);
    return saved;
  }
}

File: Frontend/src/service/storageService.js
import BaseService from '@/service/baseService';

/**
 * Service for handling storage related API operations such as
 * fetching, adding, updating, and deleting household storage items.
 * Extends the BaseService class to use pre defined HTTP methods.
 */
class StorageService extends BaseService {
  /**
   * Initializes the storage service with the '/storage' base path.
   */
  constructor() {
    super('/storage');
  }

  /**
   * Fetches all storage items for a specific household.
   *
   * @returns {Promise<Array<Object>>} Promise resolving to list of storage items.
   * @throws {Error} If the request fails.
   */
  async getStorageItemsByHousehold() {
    try {
      const response = await this.get(`household`);
      return response;
    } catch (error) {
      console.error("Error fetching storage items:", error);
      throw error;
    }
  }

  /**
   * Fetches storage items for a household filtered by item type.
   *
   * @param {string} itemType  The type of items to fetch (e.g. FOOD, LIQUIDS).
   * @returns {Promise<Array<Object>>} Promise resolving to filtered storage items.
   * @throws {Error} If the request fails.
   */
  async getStorageItemsByType(itemType) {
    try {
      const response = await this.get(`household/type/${itemType}`);
      return response;
    } catch (error) {
      console.error(`Error fetching ${itemType} items:`, error);
      throw error;
    }
  }

  /**
   * Fetches items that expire before a specified date.
   *
   * @param {Date} beforeDate  Date before which items expire.
   * @returns {Promise<Array<Object>>} Promise resolving to expiring items.
   * @throws {Error} If the request fails.
   */
  async getExpiringItems(beforeDate) {
    try {

      const formattedDate = beforeDate.toISOString();
      const response = await this.get(`household/expiring?before=${formattedDate}`);
      return response;
    } catch (error) {
      console.error("Error fetching expiring items:", error);
      throw error;
    }
  }

  /**
   * Adds a new item to the household's storage.
   *
   * @param {string} householdId  The ID of the household.
   * @param {string} itemId  The ID of the item to add.
   * @param {{ unit: string, amount: number, expirationDate?: string }} data  Item data.
   * @returns {Promise<Object>} Promise resolving to the added item response.
   * @throws {Error} If the request fails.
   */
  async addItemToStorage(itemId, data) {
    try {

      let formattedDate = null;
      if (data.expirationDate) {
        const date = new Date(data.expirationDate);
        const year = date.getFullYear();
        const month = String(date.getMonth() + 1).padStart(2, '0');
        const day = String(date.getDate()).padStart(2, '0');
        formattedDate = `${year}-${month}-${day}T00:00:00`;
      }

      const payload = {
        unit: data.unit,
        amount: data.amount,
        expirationDate: formattedDate
      };

      const response = await this.post(`household/item/${itemId}`, payload);
      return response;
    } catch (error) {
      console.error("Error adding item to storage:", error);
      throw error;
    }
  }

  /**
   * Removes an item from storage.
   *
   * @param {string} storageItemId  ID of the storage item to remove.
   * @returns {Promise<Object>} Promise resolving to the delete response.
   * @throws {Error} If the request fails.
   */
  async removeItemFromStorage(storageItemId) {
    try {
      const response = await this.post(`${storageItemId}`);
      return response;
    } catch (error) {
      console.error("Error removing item from storage:", error);
      throw error;
    }
  }

    /**
   * Updates an existing storage item with new data.
   *
   * @param {string} storageItemId  ID of the storage item to update.
   * @param {{ unit: string, amount: number, expirationDate: string }} data  Updated item data.
   * @returns {Promise<Object>} Promise resolving to the updated item response.
   * @throws {Error} If the request fails.
   */
  async updateStorageItem(storageItemId, data) {
    try {
      const payload = {
        unit: data.unit,
        amount: data.amount,
        expirationDate: data.expirationDate
      };

      const response = await this.put(`${storageItemId}`, payload);
      return response;
    } catch (error) {
      console.error("Error updating storage item:", error);
      throw error;
    }
  }
}

export default new StorageService();
